Packages/com.unity.inputsystem/InputSystem/Controls/InputControl.cs

```diff
@@ -853,6 +854,98 @@ protected internal uint stateOffsetRelativeToDeviceRoot
+        /// If then Vector3Control has 3 AxisControls as consecutive floats in memory,
+        /// then we can cast Vector3Control state memory directly to Vector3 without calling ReadUnprocessedValueFromState on x/y/z axes.
+        ///
+        /// Value is computed by automatically calling <see cref="InputControl.CalculateOptimizedControlDataType"/> when system believes control changes setup configuration.
+        /// Value can be updated manually by calling <see cref="InputControl.ApplyParameterChanges"/>.
+        /// </remarks>
+        public FourCC optimizedControlDataType => m_OptimizedControlDataType;
+
+        /// <summary>
+        /// Calculate and return a type that can represent control value in memory directly.
+        /// Value then is cached in <see cref="InputControl.optimizedControlDataType"/>.
+        /// Function shouldn't be called manually.
+        /// </summary>
+        protected virtual FourCC CalculateOptimizedControlDataType()
+        {
+            return InputStateBlock.kFormatInvalid;
+        }
+
+        /// <summary>
+        /// Apply built-in parameters changes (e.g. <see cref="AxisControl.invert"/>, others) and recompute <see cref="InputControl.optimizedControlDataType"/> for impacted controls.
+        /// </summary>
+        /// <remarks>
+        /// </remarks>
+        public void ApplyParameterChanges()
+        {
+            // First we go through all children of our own hierarchy
+            SetOptimizedControlDataTypeRecursively();
+
+            // Then we go through all parents up to the root, because our own change might influence their optimization status
+            // e.g. let's say we have a tree where root is Vector3 and children are three AxisControl
+            // And user is calling this method on AxisControl which goes from Float to NotOptimized.
+            // Then we need to also transition Vector3 to NotOptimized as well.
+
+            var currentParent = parent;
+            while (currentParent != null)
+            {
+                currentParent.SetOptimizedControlDataType();
+                currentParent = currentParent.parent;
+            }
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private void SetOptimizedControlDataType()
+        {
+            // setting check need to be inline so we clear optimizations if setting is disabled after the fact
+            m_OptimizedControlDataType = InputSettings.optimizedControlsFeatureEnabled
+                ? CalculateOptimizedControlDataType()
+                : (FourCC)InputStateBlock.kFormatInvalid;
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        internal void SetOptimizedControlDataTypeRecursively()
+        {
+            // Need to go depth-first because CalculateOptimizedControlDataType might depend on computed values of children
+            if (m_ChildCount > 0)
```

Comment:
Would recommend adding braces to this if to avoid later mistakes on what is contained in it when doing multi-line statement

Reply:
Fixed